engine/src/main/java/dgca/verifier/app/engine/DefaultCertLogicEngine.kt

```diff
@@ -67,48 +67,66 @@ class DefaultCertLogicEngine(
-                        schemaVersion
-                    ) ->
-                        when (jsonLogicValidator.isDataValid(
-                            rule.logic,
-                            dataJsonNode
-                        )) {
-                            true -> Result.PASSED
-                            false -> Result.FAIL
-                            else -> Result.OPEN
-                        }
-                    else -> Result.OPEN
+        if (rules.isEmpty()) return emptyList()
+
+        val dataJsonNode = prepareData(externalParameter, payload)
+        val hcertVersion: Triple<Int, Int, Int>? = hcertVersionString.toVersion()
+
+        return rules.map {
+            checkRule(
+                rule = it,
+                dataJsonNode = dataJsonNode,
+                hcertVersion = hcertVersion,
+                certificateType = certificateType
+            )
+        }
+    }
+
+    private fun checkRule(
+        rule: Rule,
+        dataJsonNode: ObjectNode,
+        hcertVersion: Triple<Int, Int, Int>?,
+        certificateType: CertificateType
+    ): ValidationResult {
+        val ruleEngineVersion = rule.engineVersion.toVersion()
+        val schemaVersion = rule.schemaVersion.toVersion()
+
+        val validationErrors = mutableListOf<Exception>()
+
+        val isCompatibleVersion = rule.engine == CERTLOGIC_KEY &&
+                ruleEngineVersion != null &&
+                CERTLOGIC_VERSION.isGreaterOrEqualThan(ruleEngineVersion) &&
+                hcertVersion != null &&
+                schemaVersion != null &&
+                hcertVersion.first == schemaVersion.first &&
+                hcertVersion.isGreaterOrEqualThan(schemaVersion)
+
+        val res = if (isCompatibleVersion) {
+            try {
+                when (jsonLogicValidator.isDataValid(rule.logic, dataJsonNode)) {
+                    true -> Result.PASSED
+                    false -> Result.FAIL
                 }
-                val cur: String = affectedFieldsDataRetriever.getAffectedFieldsData(
-                    rule,
-                    dataJsonNode,
-                    certificateType
-                )
-                validationResults.add(
-                    ValidationResult(
-                        rule,
-                        res,
-                        cur,
-                        null
-                    )
-                )
+            } catch (e: Exception) {
+                validationErrors.add(e)
+                Result.OPEN
             }
-            validationResults
         } else {
-            emptyList()
+            Result.OPEN
         }
 
+        val cur: String = affectedFieldsDataRetriever.getAffectedFieldsData(
+            rule,
+            dataJsonNode,
+            certificateType
+        )
+
+        return ValidationResult(
+            rule,
+            res,
+            cur,
+            if (validationErrors.isEmpty()) null else validationErrors
```

Comment:
I think this could also just stay an empty list, but I didn't want to alter the behavior to the outside, e.g. if a library consumer checks for `null`.